text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,130 @@
+
+(Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years and we have no plans to do so.)
+
+**Allow for multiple levels of package stability within the same release.**
+Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release may have different levels of stability. This means that an implementation wishing to release stable tracing today must ensure that experimental metrics are factored out in such a way that breaking changes to metrics API do not destabilize the trace API packages.
+
+## Relevant architecture
+![drawing](img/0143_cross_cutting.png)
+
+At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three separate signals. Signals share a common subsystem – context propagation – but they function independently from each other.
+
+Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principlie – seperation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon these cross-cutting APIs.
+
+OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:
+
+**API -** API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.
```

Comment:
> every signal has its own, independent API package

Is this a requirement to have separate package per signal? In .NET (and I think other languages), there is one API package and one SDK package, comprising of traces/metrics/context/baggage. We do not produce a package consisting of just one of these signal.

Reply:
So, the goal is to offer stability in one signal while allowing others to make breaking changes. In most languages, that cannot be done safely if experimental and stable code are in the same package. 

Beyond the various language-specific requirements, like ABI compatibility, I am concerned that mixing experimental and stable code will make it difficult for maintainers and contributors to reason about the effects of the changes they are making, and raises the possibility that an accidental compatibility breakage will occur. A bright line, like separate packages for each signal, makes evaluating changes simpler. 

There is also the issue of end users understanding what is stable and what is not. I do agree that most languages will want to offer a unified API for convenience sake. (I refer to it as a "global api" in this doc, which I admit may not be the best term.) If that API were just a facade for signal-specific APIs, that would make it easier to "hide" experimental signals by leaving them out of the facade, and remind the user that they are experimental by forcing them to require the extra package. 

But again, the goal here is clarity about compatibility requirements. We have to provide an implementation proposal that could reasonably achieve these goals in any language, so that implementors can get a handle on what a reasonable layout and procedure to meet those would look like. 

If .NET needs to deviate in its implementation in order to adhere to language-specific requirements and conventions, that is completely understandable. That said, is it realistic for .NET to provide clarity and compatibility when stable and experimental code is mixed together?

Reply:
Okay - my main question was, once all signals are stable, do we *require* to ship independent package for each of them? i.e instead of OpenTelemetry.API ("global"), should we split and ship OpenTelemetry.API.Metrics, OpenTelemetry.API.Traces, OpenTelemetry.API.Propagators and so on. ? 

OR,
the requirement is only to split stable vs non-stable signals into separate package?
i.e I can have OpenTelemetry.API package consisting of *all* stable signals like Baggage, Context, Traces, and separate packages for each non-stable signal, like OpenTelemetry.API.Metric, OpenTelemetry.API.Logs, each of these shipping as pre-release, or with package names ending in .experimental?

Reply:
> That said, is it realistic for .NET to provide clarity and compatibility when stable and experimental code is mixed together?

The way .NET packaging is done currently is Metrics are part of the same package as Traces. And we marked all Metric methods/classes as "Obsolete", which would sufficiently warn users (compile time warnings) that it is experimental as shown here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/src/OpenTelemetry/Metrics/MeterProviderBuilder.cs#L28-L29

(We are still open to splitting metrics into separate package, that's being discussed in .NET sig)

Reply:
Really, the goal is just ensuring stability and clearly communicating that stability to end users. If that can be done without splitting up the API, I don't see a problem from a versioning and stability perspective. But, would that not create ABI compatibilities? Or is that not an issue in .NET?

Personally, marking beta code as OBSOLETE sounds a hack, but if that is a reasonable practice in .NET I would not take any issue with it. If users are going to respond with a "wtf" I'd suggest taking a different approach.

Reply:
> Personally, marking beta code as OBSOLETE sounds a hack,

Yes. Its can be called hack. But it surely gets people attention, as it generates compile time warnings.
I wish there was an attribute called "Experimental", just like "Obsolete"!

In .NET versioning proposal, we have listed both approaches. Leaving it open to get more feedback from  .NET SIG, but it looks like we are leaning towards separate package for unstable signals. (that aligns more closely with other SIGs as well)

Reply:
Thank you, really appreciate the thought you are putting into this.

Reply:
(I also wish our programming languages would do more to explicitly support stability and change management, OBSOLETE is a good start)